icu4c/source/i18n/rbt_pars.cpp

```diff
@@ -557,6 +557,7 @@ int32_t RuleHalf::parseSection(const UnicodeString& rule, int32_t pos, int32_t l
                 // The next character MUST be a segment open
                 if (single == NULL ||
                     !ICU_Utility::parseChar(rule, iref, SEGMENT_OPEN)) {
+                    delete single;
```

Comment:
could we change 
```
TransliteratorIDParser::SingleID* single =
                    TransliteratorIDParser::parseFilterID(rule, iref);
```
to
```
LocalMemory<TransliteratorIDParser::SingleID> single(
                    TransliteratorIDParser::parseFilterID(rule, iref));
```
to fix this?

Reply:
LocalPointer, not LocalMemory

Then
```
if (single.isNull() || !ICU_Util...()) {
    return syntaxError(...);
}
```
Following code also would no longer `delete single`.

Reply:
@FrankYFTang like this?:
```
diff --git a/icu4c/source/i18n/rbt_pars.cpp b/icu4c/source/i18n/rbt_pars.cpp
index 2c2fd3a414f..b2ced2f1926 100644
--- a/icu4c/source/i18n/rbt_pars.cpp
+++ b/icu4c/source/i18n/rbt_pars.cpp
@@ -552,17 +552,15 @@ int32_t RuleHalf::parseSection(const UnicodeString& rule, int32_t pos, int32_t l
         case ALT_FUNCTION:
             {
                 int32_t iref = pos;
-                TransliteratorIDParser::SingleID* single =
-                    TransliteratorIDParser::parseFilterID(rule, iref);
+                LocalPointer<TransliteratorIDParser::SingleID> single(
+                    TransliteratorIDParser::parseFilterID(rule, iref));
                 // The next character MUST be a segment open
-                if (single == nullptr ||
+                if (single.isNull() ||
                     !ICU_Utility::parseChar(rule, iref, SEGMENT_OPEN)) {
-                    delete single;
                     return syntaxError(U_INVALID_FUNCTION, rule, start, status);
                 }
                 
                 Transliterator *t = single->createInstance();
-                delete single;
                 if (t == nullptr) {
                     return syntaxError(U_INVALID_FUNCTION, rule, start, status);
                 }
```